lib/path-utils.js

```diff
@@ -0,0 +1,225 @@
+    // So we trim the slash to make it relative to project folder from glob perspective.
+    entry = entry.substring(1)
+    // Check if it is a directory or file
+    if (await isPath(entry)) {
+      pathType = await getPathType(gitIgnoreDirectory ? path.join(gitIgnoreDirectory, entry) : entry)
+    }
+  } else {
+    const slashPlacement = entry.indexOf("/")
+    if (slashPlacement === -1) {
+      // Patterns that don't have `/` are '**/' from glob perspective (can match at any level)
+      if (!entry.startsWith("**/")) {
+        entry = `**/${entry}`
+      }
+    } else if (slashPlacement === entry.length - 1) {
+      // If there is a separator at the end of the pattern then it only matches directories
+      // slash is in the end
+      pathType = 1
+    } else {
+      // has `/` in the middle so it is a relative path
+      // Check if it is a directory or file
+      if (await isPath(entry)) {
+        pathType = await getPathType(gitIgnoreDirectory ? path.join(gitIgnoreDirectory, entry) : entry)
+      }
+    }
+  }
+
+  // prepend the absolute root directory
+  if (gitIgnoreDirectory) {
+    entry = `${posixifyPath(gitIgnoreDirectory)}/${entry}`
+  }
+
+  // swap !
+  entry = forceInclude ? entry : `!${entry}`
+
+  // Process the entry ending
+  if (/** is directory */ pathType === 1) {
+    // in glob this is equal to `directry/**`
+    if (entry.endsWith("/")) {
+      return `${entry}**`
+    } else {
+      return `${entry}/**`
+    }
+  } else if (/** is file */ pathType === 2) {
+    return entry
+  } else if (!entry.endsWith("/**")) {
+    // the pattern can match both files and directories
+    // so we should include both `entry` and `entry/**`
+    return [entry, `${entry}/**`]
+  } else {
+    return entry
+  }
+}
+
+function isWhitespace(str) {
+  return /^\s*$/.test(str)
+}
+
+/**
+ * A line starting with # serves as a comment.
+ * Put a backslash ("\") in front of the first hash for patterns that begin with a hash.
+ */
+function isGitIgnoreComment(pattern) {
+  return pattern[0] === "#"
+}
+
+/**
+ * Trailing spaces should be removed unless they are quoted with backslash ("\ ").
+ */
+function trimTrailingWhitespace(str) {
+  if (!/\\\s+$/.test(str)) {
+    // No escaped trailing whitespace, remove
+    return str.replace(/\s+$/, "")
+  } else {
+    // Trailing whitespace detected, remove only the backslash
+    return str.replace(/\\(\s+)$/, "$1")
+  }
+}
+
+/** Remove leading whitespace */
+function trimLeadingWhiteSpace(str) {
+  return str.replace(/^\s+/, "")
+}
+
+/** Remove whitespace from a gitignore entry */
+function trimWhiteSpace(str) {
+  return trimLeadingWhiteSpace(trimTrailingWhitespace(str))
+}
+
+/** Get the type of the given path (1 for directory, 2 for file, 0 for others)
+ * @param {string} givenPath absolute path
+ * @returns {Promise<number>}
+ */
+async function getPathType(filepath) {
+  let pathStat
+  try {
+    pathStat = await stat(filepath)
+  } catch (error) {
+    return 0
```

Comment:
Maybe we should make an enum.

```js
const PATH_TYPE = {
	OTHER: 0,
	DIRECTORY: 1,
	FILE: 2,
}
```

Reply:
Fixed in https://github.com/atom-community/autocomplete-paths/pull/256/commits/9b35b4fa8ec555ec49dc77c7b481929fff4e44dd